types/block.go

```diff
@@ -741,6 +740,160 @@ func (cs *CommitSig) FromProto(csp tmproto.CommitSig) error {
 
 //-------------------------------------
 
+// ExtendedCommitSig is similar to CommitSig, except it also contains the vote
+// extension and vote extension signature related to this commit.
+type ExtendedCommitSig struct {
+	BlockIDFlag        BlockIDFlag `json:"block_id_flag"`
+	ValidatorAddress   Address     `json:"validator_address"`
+	Timestamp          time.Time   `json:"timestamp"`
+	Signature          []byte      `json:"signature"`
+	VoteExtension      []byte      `json:"extension"`
+	ExtensionSignature []byte      `json:"extension_signature"`
+}
+
+// NewExtendedCommitSigAbsent returns new ExtendedCommitSig with
+// BlockIDFlagAbsent. Other fields are all empty.
+// TODO(sergio): When all UTs passing, try removing the Commit and CommitSig version of new functions
+func NewExtendedCommitSigAbsent() ExtendedCommitSig {
+	return ExtendedCommitSig{
+		BlockIDFlag: BlockIDFlagAbsent,
+	}
+}
+
+// ForBlock returns true if ExtendedCommitSig is for the block.
+func (ecs ExtendedCommitSig) ForBlock() bool {
+	return ecs.BlockIDFlag == BlockIDFlagCommit
+}
+
+// Absent returns true if ExtendedCommitSig is absent.
+func (ecs ExtendedCommitSig) Absent() bool {
+	//TODO What about BlockIDFlagNil?
+	return ecs.BlockIDFlag == BlockIDFlagAbsent
+}
+
+// String returns a string representation of an ExtendedCommitSig.
+//
+// 1. first 6 bytes of signature
+// 2. first 6 bytes of validator address
+// 3. block ID flag
+// 4. timestamp
+// 5. first 6 bytes of vote extension
+// 6. first 6 bytes of vote extension signature
+func (ecs ExtendedCommitSig) String() string {
+	return fmt.Sprintf("ExtendedCommitSig{%X by %X on %v @ %s with %X %X}",
+		tmbytes.Fingerprint(ecs.Signature),
+		tmbytes.Fingerprint(ecs.ValidatorAddress),
+		ecs.BlockIDFlag,
+		CanonicalTime(ecs.Timestamp),
+		tmbytes.Fingerprint(ecs.VoteExtension),
+		tmbytes.Fingerprint(ecs.ExtensionSignature),
+	)
+}
+
+func (ecs ExtendedCommitSig) BlockID(commitBlockID BlockID) BlockID {
+	var blockID BlockID
+	switch ecs.BlockIDFlag {
+	case BlockIDFlagAbsent:
+		blockID = BlockID{}
+	case BlockIDFlagCommit:
+		blockID = commitBlockID
+	case BlockIDFlagNil:
+		blockID = BlockID{}
+	default:
+		panic(fmt.Sprintf("Unknown BlockIDFlag: %v", ecs.BlockIDFlag))
+	}
+	return blockID
+}
+
+// ValidateBasic checks whether the structure is well-formed.
+func (ecs ExtendedCommitSig) ValidateBasic() error {
+	switch ecs.BlockIDFlag {
+	case BlockIDFlagAbsent:
+	case BlockIDFlagCommit:
+	case BlockIDFlagNil:
+	default:
+		return fmt.Errorf("unknown BlockIDFlag: %v", ecs.BlockIDFlag)
+	}
+
+	switch ecs.BlockIDFlag {
```

Comment:
There's often a preference in Go code to early return for legibility and to stick-to-the left. Since the switch is basically an `if` statement that checks for an `Absent` BlockID value, you can transform the whole `switch` into a single clause `if` and leave all the other validation checks as no longer indented.


```go
if ecs.BlockIDFlag == BlockIDFlagAbsent { 
if len(ecs.Extension) != 0 {
return fmt.Errorf(... 
}

// here, we 'stick to the left' and no longer keep the checks for the common cases indented. 
// A little easier to read without as many indents (imo).
if len(ecs.Extension) > MaxVoteExtensionSize {
return fmt.Errorf(...
}
...

return nil
```

Reply:
Ah I see, I've refactored this function now.